R/estimateDiversity.R

```diff
@@ -178,13 +188,123 @@ setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
+                       "coverage", "fisher", "faith"),
+             name = index, ..., BPPARAM = SerialParam()){
+        
+        # Gets the tree 
+        tree <- rowTree(x)
+        
+        # input check
+        # If object does not have a tree
+        if( ("faith" %in% index) && (is.null(tree) || is.null(tree$edge.length)) ){
+            stop("Object does not have a tree or the tree does not have any branches.
+             'faith' is not possible to calculate.",
+                 call. = FALSE)
+        }
+        
+        index<- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                 "same length than 'index'.",
+                 call. = FALSE)
+        }
+        
+        # If 'faith' is one of the indices
+        if( "faith" %in% index ){
+            # Get the name of "faith" index
+            faith_name <- name[index %in% "faith"]
+            # And delete it from name
+            name <- name[!index %in% "faith"]
+
+            # Delete "faith" from indices
+            index <- index[!index %in% "faith"]
+            
+            # Faith will be calculated
+            calc_faith <- TRUE
+        } else{
+            # Faith will not be calculated
+            calc_faith <- FALSE
+        }
+
+        # If index list contained other than 'faith' index, the length of the list is over 0
+        if( length(index)>0){
+            # Calculates all indices but not 'faith'
+            x <- callNextMethod()
+        }
+
+        # If 'faith' was one of the indices, 'calc_faith' is TRUE
+        if( calc_faith ){
+            # Calculates faith
+            x <- estimateFaith(x, tree = tree, name = faith_name, ...)
+        }
 
+        return(x)
+
+    }
+)
+
+#' @rdname estimateDiversity
+#' 
+#' @param tree A phylogenetic tree that is used to calculate 'faith' index.
+#' By default, it is \code{rowTree(x)}.
+#' 
+#' @export
+setGeneric("estimateFaith",signature = c("x", "tree"),
+           function(x, tree = "missing", abund_values = "counts",
+                    name = "faith", ...)
+               standardGeneric("estimateFaith"))
+
+#' @rdname estimateDiversity
+#' @export
+setMethod("estimateFaith", signature = c(x="SummarizedExperiment", tree="phylo"),
+    function(x, tree, abund_values = "counts",
+            name = "faith", ...){
+
+        # Input check
+        # Check 'abund_values'
+        .check_assay_present(abund_values, x)
+
+        # Calculates Faith index
+        faith <- list(.calc_faith(assay(x, abund_values), tree))
+
+        # Adds calculated Faith index to colData
+        .add_values_to_colData(x, faith, name)
+
+          }
+)
+
+#' @rdname estimateDiversity
+#' @export
+setMethod("estimateFaith", signature = c(x="TreeSummarizedExperiment", tree="missing"),
+    function(x, tree = rowTree(x), abund_values = "counts",
```

Comment:
The same thing as for you most recent changes applies here. If you remove the tree argument from the function definition alltogether, which you are allowed to do, since `tree` is defined as missing, and add `tree <- rowTree(x )` in the function it will be more consistent.

The rational behind it is as follow:
1. SE + tree: dispatch directly to correct method
2. TSE: tree gets extracted and next call to `estimateFaith` is issued. Since a TSE is also a SE, it matches the signature for 1.
1. If run with a TSE + tree, it gets dispatched to 1. directly as done internally in 2. This also works it the TSE does not contain a `rowTree`.

This is the beauty of S4 dispatch 😄 to cover all the possible combinations of available data

Reply:
Thanks for clarifying, I think that I got it